src/test/blake2s_test.cpp

```diff
@@ -0,0 +1,234 @@
+
+    libsnark::pb_variable_array<FieldT> c2;
+    c2.allocate(pb, 32, "c2");
+
+    libsnark::pb_variable_array<FieldT> d2;
+    d2.allocate(pb, 32, "d2");
+
+    g_primitive<FieldT> g_gadget(pb, a, b, c, d, x, y, a2, b2, c2, d2);
+    g_gadget.generate_r1cs_constraints();
+    g_gadget.generate_r1cs_witness();
+
+    libsnark::pb_variable_array<FieldT> a2_expected = from_bits(
+        {
+            0, 1, 1, 1, 0, 0, 0, 0, // 70
+            1, 0, 1, 1, 0, 0, 0, 1, // B1
+            0, 0, 1, 1, 0, 1, 0, 1, // 35
+            0, 0, 1, 1, 1, 1, 0, 1  // 3D
+        },
+        ZERO);
+
+    libsnark::pb_variable_array<FieldT> b2_expected = from_bits(
+        {
+            1, 1, 0, 0, 0, 0, 0, 0, // C0
+            0, 1, 1, 1, 1, 1, 1, 1, // 7F
+            0, 0, 1, 0, 1, 1, 1, 0, // 2E
+            0, 1, 1, 1, 1, 0, 1, 1  // 7B
+        },
+        ZERO);
+
+    libsnark::pb_variable_array<FieldT> c2_expected = from_bits(
+        {
+            1, 1, 1, 0, 0, 1, 1, 1, // E7
+            0, 0, 1, 0, 0, 0, 0, 1, // 21
+            0, 1, 0, 0, 1, 0, 1, 1, // 4B
+            0, 1, 0, 0, 0, 0, 0, 0  // 40
+        },
+        ZERO);
+
+    libsnark::pb_variable_array<FieldT> d2_expected = from_bits(
+        {
+            1, 0, 1, 1, 0, 0, 0, 0, // B0
+            1, 0, 1, 1, 1, 1, 0, 0, // BC
+            1, 1, 1, 0, 1, 0, 1, 1, // EB
+            0, 1, 0, 0, 1, 1, 0, 0  // 4C
+        },
+        ZERO);
+
+    ASSERT_EQ(a2_expected.get_bits(pb), a2.get_bits(pb));
+    ASSERT_EQ(b2_expected.get_bits(pb), b2.get_bits(pb));
+    ASSERT_EQ(c2_expected.get_bits(pb), c2.get_bits(pb));
+    ASSERT_EQ(d2_expected.get_bits(pb), d2.get_bits(pb));
+}
+
+// The test correponds to blake2s(b"hello world")
+TEST(TestBlake2sComp, TestTrue)
+{
+    libsnark::protoboard<FieldT> pb;
+
+    libsnark::pb_variable<FieldT> ZERO;
+    ZERO.allocate(pb, "zero");
+    pb.val(ZERO) = FieldT::zero();
+
+    // b"hello world" in big endian
+    libsnark::pb_variable_array<FieldT> pb_va_input = from_bits(
+        {
+            0, 1, 1, 0, 1, 0, 0, 0, // 68
+            0, 1, 1, 0, 0, 1, 0, 1, // 65
+            0, 1, 1, 0, 1, 1, 0, 0, // 6C
+            0, 1, 1, 0, 1, 1, 0, 0, // 6C
+            0, 1, 1, 0, 1, 1, 1, 1, // 6F
+            0, 0, 1, 0, 0, 0, 0, 0, // 20
+            0, 1, 1, 1, 0, 1, 1, 1, // 77
+            0, 1, 1, 0, 1, 1, 1, 1, // 6F
+            0, 1, 1, 1, 0, 0, 1, 0, // 72
+            0, 1, 1, 0, 1, 1, 0, 0, // 6C
+            0, 1, 1, 0, 0, 1, 0, 0, // 64
+        },
+        ZERO);
+
+    libsnark::block_variable<FieldT> input(
+        pb, {pb_va_input}, "blake2s_block_input");
+
+    libsnark::digest_variable<FieldT> output(pb, BLAKE2s_digest_size, "output");
+
+    BLAKE2s_256_comp<FieldT> blake2s_comp_gadget(pb, input, output);
+    blake2s_comp_gadget.generate_r1cs_constraints();
+    blake2s_comp_gadget.generate_r1cs_witness();
+
+    // blake2s(b"hello world")
+    libsnark::pb_variable_array<FieldT> expected = from_bits(
```

Comment:
What did you use to generate the test vectors @rrtoledo?

Reply:
I used hashlib for `TestBlake2sComp`.
As for `TestG`, I used a python script (and added prints) referred by blake2.net (https://github.com/buggywhip/blake2_py) .

Reply:
Can you add a line to specify this in the code so it's clear where these values come from?

Reply:
Done

Reply:
Thanks